lib/babe/epoch.go

```diff
@@ -221,41 +222,71 @@ func (b *Service) incrementEpoch() (uint64, error) {
 		epochData.randomness,
-		slot,
-		epoch,
+		slotNumber,
+		epochNumber,
 		epochData.threshold,
-		b.keypair,
+		keypair,
 	)
-	if err == nil {
-		preRuntimeDigest := types.NewBabePrimaryPreDigest(epochData.authorityIndex, slot, proof.output, proof.proof)
 
-		return preRuntimeDigest.ToPreRuntimeDigest()
+	if err == nil {
+		preRuntimeDigest, err := types.ToPreRuntimeDigest(*types.NewBabePrimaryPreDigest(
+			epochData.authorityIndex, slotNumber, proof.output, proof.proof))
+		if err != nil {
+			return nil, fmt.Errorf("error converting babe primary pre-digest to pre-runtime digest: %w", err)
+		}
+		logger.Debugf("epoch %d: claimed primary slot %d", epochNumber, slotNumber)
+		return preRuntimeDigest, nil
+	} else if !errors.Is(err, errOverPrimarySlotThreshold) {
+		return nil, fmt.Errorf("error running slot lottery at slot %d: %w", slotNumber, err)
 	}
 
-	if errors.Is(err, errOverPrimarySlotThreshold) {
-		if epochData.secondary == 0 {
-			return nil, errSecondarySlotProductionDisabled
+	switch epochData.secondary {
+	case types.PrimarySlots:
+		return nil, errNotOurTurnToPropose
+	case types.PrimaryAndSecondaryVRFSlots:
+		proof, err := claimSecondarySlotVRF(
+			epochData.randomness, slotNumber, epochNumber, epochData.authorities, keypair, epochData.authorityIndex)
+		if err != nil {
+			return nil, fmt.Errorf("error claim secondary vrf slot at %d: %w", slotNumber, err)
 		}
 
-		proof, err = claimSecondarySlot(epochData.randomness, slot, epoch,
-			epochData.authorities, b.keypair, epochData.authorityIndex)
+		preRuntimeDigest, err := types.ToPreRuntimeDigest(*types.NewBabeSecondaryVRFPreDigest(
+			epochData.authorityIndex, slotNumber, proof.output, proof.proof))
 		if err != nil {
-			return nil, fmt.Errorf("cannot claim secondary slot: %w", err)
+			return nil, fmt.Errorf("error converting babe secondary vrf pre-digest to pre-runtime digest: %w", err)
 		}
-		if proof != nil {
-			preRuntimeDigest := types.NewBabeSecondaryPlainPreDigest(epochData.authorityIndex, slot)
-			return preRuntimeDigest.ToPreRuntimeDigest()
+
+		logger.Debugf("epoch %d: claimed secondary vrf slot %d", epochNumber, slotNumber)
+		return preRuntimeDigest, nil
+	case types.PrimaryAndSecondaryPlainSlots:
+		err = claimSecondarySlotPlain(
+			epochData.randomness, slotNumber, epochData.authorities, epochData.authorityIndex)
+		if err != nil {
+			return nil, fmt.Errorf("error claiming secondary plain slot at %d: %w", slotNumber, err)
+		}
+
+		preRuntimeDigest, err := types.ToPreRuntimeDigest(*types.NewBabeSecondaryPlainPreDigest(
+			epochData.authorityIndex, slotNumber))
+
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to get preruntime digest from babe secondary plain predigest for slot %d: %w", slotNumber, err)
 		}
-	}
 
-	return nil, err
+		logger.Debugf("epoch %d: claimed secondary plain slot %d", epochNumber, slotNumber)
+		return preRuntimeDigest, nil
+	default:
+		// this should never occur
+		return nil, errInvalidSlotTechnique
```

Comment:
> Although if it's fully tested, there shouldn't be a need for this case nor the panic really

I would rather have such dead code than not, just in case! Helps, cover all cases!